jobspec2/parse.go

```diff
@@ -14,75 +16,99 @@ import (
 	hcljson "github.com/hashicorp/hcl/v2/json"
 	"github.com/hashicorp/nomad/api"
 	"github.com/hashicorp/nomad/jobspec2/hclutil"
-	"github.com/zclconf/go-cty/cty"
 )
 
 func Parse(path string, r io.Reader) (*api.Job, error) {
-	return ParseWithArgs(path, r, nil, false)
-}
-
-func toVars(vars map[string]string) cty.Value {
-	attrs := make(map[string]cty.Value, len(vars))
-	for k, v := range vars {
-		attrs[k] = cty.StringVal(v)
-	}
-
-	return cty.ObjectVal(attrs)
-}
-
-func ParseWithArgs(path string, r io.Reader, vars map[string]string, allowFS bool) (*api.Job, error) {
 	if path == "" {
 		if f, ok := r.(*os.File); ok {
 			path = f.Name()
 		}
 	}
-	basedir := filepath.Dir(path)
 
-	// Copy the reader into an in-memory buffer first since HCL requires it.
 	var buf bytes.Buffer
-	if _, err := io.Copy(&buf, r); err != nil {
+	_, err := io.Copy(&buf, r)
+	if err != nil {
 		return nil, err
 	}
 
-	evalContext := &hcl.EvalContext{
-		Functions: Functions(basedir, allowFS),
-		Variables: map[string]cty.Value{
-			"vars": toVars(vars),
-		},
-		UnknownVariable: func(expr string) (cty.Value, error) {
-			v := "${" + expr + "}"
-			return cty.StringVal(v), nil
-		},
-	}
-	var result struct {
-		Job jobWrapper `hcl:"job,block"`
+	return ParseWithConfig(&ParseConfig{
+		Path:    path,
+		Body:    buf.Bytes(),
+		AllowFS: false,
+		Strict:  true,
+	})
+}
+
+func ParseWithConfig(args *ParseConfig) (*api.Job, error) {
+	args.normalize()
+
+	c := &jobConfig{
```

Comment:
It doesn't look like we construct this with its `InputVariables` and `LocalVariables` instantiated, which are maps. I think we end up handling it safely because of how the maps are wrapped by the `Variables` alias but it might be safer for future use to have a `newJobConfig` that sets everything up safely local to where it's declared?

Reply:
good point - done in c2a6a6da8063919dcfcc8eac1e82549791604b22